geo/src/main/java/org/odk/collect/geo/geopoly/GeoPolyFragment.kt

```diff
@@ -0,0 +1,631 @@
+        this.accuracyThresholdIndex = accuracyThresholdIndex
+    }
+
+    private fun cancel() {
+        getParentFragmentManager().setFragmentResult(REQUEST_GEOPOLY, Bundle.EMPTY)
+    }
+
+    private fun onClick(point: MapPoint) {
+        if (inputActive && !recordingEnabled) {
+            appendPointIfNew(point)
+        }
+    }
+
+    private fun onGpsLocationReady(map: MapFragment) {
+        // Don't zoom to current location if a user is manually entering points
+        if (requireActivity().window.isActive && (!inputActive || recordingEnabled)) {
+            map.zoomToCurrentLocation(map.getGpsLocation())
+        }
+        updateUi()
+    }
+
+    private fun onGpsLocation(point: MapPoint?) {
+        if (inputActive && recordingEnabled) {
+            map!!.setCenter(point, false)
+        }
+        updateUi()
+    }
+
+    private fun recordPoint(point: MapPoint?) {
+        if (point != null && isLocationAcceptable(point)) {
+            appendPointIfNew(point)
+        }
+    }
+
+    private fun appendPointIfNew(point: MapPoint) {
+        val points = map!!.getPolyLinePoints(featureId)
+        if (points.isEmpty() || point != points[points.size - 1]) {
+            map!!.appendPointToPolyLine(featureId, point)
+            updateUi()
+        }
+
+        setResult(RESULT_GEOPOLY_CHANGE)
+    }
+
+    private fun isLocationAcceptable(point: MapPoint): Boolean {
+        if (!this.isAccuracyThresholdActive) {
+            return true
+        }
+        return point.accuracy <= ACCURACY_THRESHOLD_OPTIONS[accuracyThresholdIndex]
+    }
+
+    private val isAccuracyThresholdActive: Boolean
+        get() {
+            val meters: Int =
+                ACCURACY_THRESHOLD_OPTIONS[accuracyThresholdIndex]
+            return recordingEnabled && recordingAutomatic && meters > 0
+        }
+
+    private fun removeLastPoint() {
+        if (featureId != -1) {
+            map!!.removePolyLineLastPoint(featureId)
+            updateUi()
+            setResult(RESULT_GEOPOLY_CHANGE)
+        }
+    }
+
+    private fun clear() {
+        map!!.clearFeatures()
+        featureId = map!!.addPolyLine(
+            LineDescription(
+                emptyList(),
+                MapConsts.DEFAULT_STROKE_WIDTH.toString(),
+                null,
+                !readOnly,
+                outputMode == OutputMode.GEOSHAPE
+            )
+        )
+        inputActive = false
+        updateUi()
+    }
+
+    /** Updates the state of various UI widgets to reflect internal state.  */
+    private fun updateUi() {
+        val numPoints = map!!.getPolyLinePoints(featureId).size
+        val location = map!!.getGpsLocation()
+
+        // Visibility state
+        playButton!!.visibility = if (inputActive) View.GONE else View.VISIBLE
+        pauseButton!!.visibility = if (inputActive) View.VISIBLE else View.GONE
+        recordButton!!.visibility = if (inputActive && recordingEnabled && !recordingAutomatic) View.VISIBLE else View.GONE
```

Comment:
Here and above you can use `isVisible = inputActive && recordingEnabled && !recordingAutomatic` so that the expression is simpler as you don't need the if statement.